tools/autograd/templates/variable_factories.h

```diff
@@ -17,6 +18,125 @@
 using at::DimnameList;
 #endif
 
+namespace {
+  enum class ListInitTensorType { Scalar, InitList };
+
+  // We use `ListInitTensor` to support converting an arbitrarily nested braced-init-list
+  // (e.g. {{1, 2}, {3, 4}}) into the equivalent Tensor, taking advantage of the fact that
+  // the constructor will automatically be called recursively until it reaches all innermost
+  // scalar values.
+  //
+  // At any time, a `ListInitTensor` object represents either of the following:
+  // 1. A scalar with value `scalar()` and type `scalar_type()`.
+  // 2. A Tensor represented in `std::initializer_list<ListInitTensor>` form, with value
+  //    `init_list()`, Tensor scalar type `scalar_type()`, and Tensor sizes `sizes()`.
+  struct ListInitTensor {
+#define TENSOR(T, S)                   \
+    ListInitTensor(T scalar) :         \
+        scalar_(scalar), init_list_(), \
+        sizes_(),                      \
+        scalar_type_(at::k##S),        \
+        type_(ListInitTensorType::Scalar) {}
+AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
+#undef TENSOR
+    ListInitTensor(std::initializer_list<ListInitTensor> init_list) :
+        scalar_(),
+        init_list_(init_list),
+        sizes_(),
+        scalar_type_(init_list.begin()->scalar_type_),
+        type_(ListInitTensorType::InitList) {
+      const ListInitTensor& first_elem = *(init_list.begin());
```

Comment:
This assumes that the list has at least one element. Is that true?

Reply:
I added a check to enforce that the init-lists are not empty. Passing empty init-lists also causes ambiguous overload error:
```
../test/cpp/api/tensor.cpp:201:41: error: call of overloaded ‘tensor(<brace-enclosed initializer list>)’ is ambiguous
     auto tensor = torch::tensor({{}, {}});
                                         ^
In file included from ../caffe2/../torch/csrc/api/include/torch/types.h:7:0,
                 from ../test/cpp/api/tensor.cpp:3:
../torch/csrc/autograd/generated/variable_factories.h:169:644: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<unsigned char>)
../torch/csrc/autograd/generated/variable_factories.h:169:766: note: candidate: at::Tensor torch::tensor(std::initializer_list<unsigned char>)
../torch/csrc/autograd/generated/variable_factories.h:169:1603: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<signed char>)
../torch/csrc/autograd/generated/variable_factories.h:169:1724: note: candidate: at::Tensor torch::tensor(std::initializer_list<signed char>)
../torch/csrc/autograd/generated/variable_factories.h:169:2562: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<short int>)
../torch/csrc/autograd/generated/variable_factories.h:169:2685: note: candidate: at::Tensor torch::tensor(std::initializer_list<short int>)
../torch/csrc/autograd/generated/variable_factories.h:169:3507: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<int>)
../torch/csrc/autograd/generated/variable_factories.h:169:3624: note: candidate: at::Tensor torch::tensor(std::initializer_list<int>)
../torch/csrc/autograd/generated/variable_factories.h:169:4450: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<long int>)
../torch/csrc/autograd/generated/variable_factories.h:169:4572: note: candidate: at::Tensor torch::tensor(std::initializer_list<long int>)
../torch/csrc/autograd/generated/variable_factories.h:169:5404: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<float>)
../torch/csrc/autograd/generated/variable_factories.h:169:5525: note: candidate: at::Tensor torch::tensor(std::initializer_list<float>)
../torch/csrc/autograd/generated/variable_factories.h:169:6354: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<double>)
../torch/csrc/autograd/generated/variable_factories.h:169:6477: note: candidate: at::Tensor torch::tensor(std::initializer_list<double>)
../torch/csrc/autograd/generated/variable_factories.h:169:7630: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<bool>)
../torch/csrc/autograd/generated/variable_factories.h:169:7815: note: candidate: at::Tensor torch::tensor(std::initializer_list<bool>)
../torch/csrc/autograd/generated/variable_factories.h:169:9224: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<c10::Half>)
../torch/csrc/autograd/generated/variable_factories.h:169:9409: note: candidate: at::Tensor torch::tensor(std::initializer_list<c10::Half>)
../torch/csrc/autograd/generated/variable_factories.h:169:9656: note: candidate: at::Tensor torch::tensor(c10::impl::ScalarTypeToCPPType<(c10::ScalarType)5>::type)
../torch/csrc/autograd/generated/variable_factories.h:169:10838: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<c10::BFloat16>)
../torch/csrc/autograd/generated/variable_factories.h:169:11031: note: candidate: at::Tensor torch::tensor(std::initializer_list<c10::BFloat16>)
../torch/csrc/autograd/generated/variable_factories.h:169:11286: note: candidate: at::Tensor torch::tensor(c10::impl::ScalarTypeToCPPType<(c10::ScalarType)15>::type)
In file included from ../caffe2/../torch/csrc/api/include/torch/types.h:7:0,
                 from ../test/cpp/api/tensor.cpp:3:
../torch/csrc/autograd/generated/variable_factories.h:176:19: note: candidate: at::Tensor torch::tensor({anonymous}::ListInitTensor)
 inline at::Tensor tensor(ListInitTensor list_init_tensor) {
                   ^
```
If the user wants to create an empty tensor, they can use `torch::randn({2, 0})` for now. And if they really want `torch::tensor({{}, {}})` we can try to find ways to fix the overload problem later.

Reply:
Yup, that sounds good! I was just checking that the code doesn't segfault